### PR TITLE
Add mesh_layout argument with rectilinear two-step coordinate/connectivity architecture

### LINKED ISSUE #78: [Feat] Introduce `mesh_layout` argument to `create_all_graph_components` (default: `rectilinear`)
## Summary

Introduce a `mesh_layout` argument to `create_all_graph_components()` that controls **how mesh nodes are spatially arranged**, decoupled from `m2m_connectivity` which controls **how mesh levels relate to each other**.

This is the foundational change that enables all subsequent layout options (`triangular`, `prebuilt`, `icosahedral`).

## Motivation

Currently, `m2m_connectivity="flat"` implies both:
- "The mesh is a single-level flat structure" (connectivity topology)
- "The mesh nodes are laid out on a rectilinear grid" (spatial layout)

These are distinct concepts. Separating them enables combinations like `m2m_connectivity="hierarchical" + mesh_layout="triangular"` — a hierarchical graph with triangular mesh at each level.

Discussion origin: #71

## Proposed API Change

```python
def create_all_graph_components(
    coords,
    m2m_connectivity,
    mesh_layout="rectilinear",     # NEW — default preserves current behavior
    mesh_layout_kwargs={},         # NEW
    m2m_connectivity_kwargs={},
    g2m_connectivity,
    m2g_connectivity,
    ...
):
```
### Implementation Plan

1. Add mesh_layout and mesh_layout_kwargs parameters to `create_all_graph_components()` in `base.py`
2. Add dispatch logic before the existing `m2m_connectivity` dispatch:
```
# Step 1: Create single-level mesh graph based on layout
if mesh_layout == "rectilinear":
    # Current behavior — no change
    create_mesh_fn = create_single_level_2d_mesh_graph
elif mesh_layout == "triangular":
    create_mesh_fn = create_single_level_2d_triangular_mesh_graph  # future
elif mesh_layout == "prebuilt":
    create_mesh_fn = _validate_and_use_prebuilt_mesh  # future
elif mesh_layout == "icosahedral":
    create_mesh_fn = create_icosahedral_mesh  # future
else:
    raise ValueError(f"Unknown mesh_layout: {mesh_layout}")

# Step 2: m2m_connectivity uses the mesh graph from step 1
# (existing logic, unchanged for "rectilinear")
```

3. For this initial PR: only mesh_layout="rectilinear" is functional — other values raise `NotImplementedError` with a message pointing to the relevant tracking issue
4. Update archetype functions in `archetype.py` to pass through mesh_layout and mesh_layout_kwargs
5. Add tests verifying that mesh_layout="rectilinear" produces identical output to the current code (no behavioral change)
6. Update docstrings to document the new parameter

### Backward Compatibility
Default mesh_layout="rectilinear" means zero change for all existing users
All existing tests pass without modification
Archetype functions (`create_keisler_graph`, `create_graphcast_graph, `create_oskarsson_hierarchical_graph`) work unchanged

### Related Issues
#71 — Original discussion (to be closed once sub-issues are created)
#70 — mesh_layout=prebuilt implementation
#75 — CRS-based distance metric auto-detection
#77 — mesh_layout=icosahedral

### PR DESCRIPTION
**Pull request by prajwal-tech07**
### Add `mesh_layout` argument with rectilinear two-step coordinate/connectivity architecture
Implements the `mesh_layout` argument for mesh graph creation, starting with `rectilinear` as the first supported layout. This uses a **two-step architecture** that separates:

1. **Coordinate creation** (`nx.Graph`) — places mesh nodes on a grid with `cardinal` and `diagonal` edge annotations
2. **Connectivity creation** (`nx.DiGraph`) — filters edges by pattern (`"4-star"` or `"8-star"`) and computes edge features

### Changes by file:

**`base.py`**
- Added `mesh_layout` (default: `"rectilinear"`) and `mesh_layout_kwargs` parameters to `create_single_level_2d_mesh_graph` and `create_multi_level_2d_mesh_graph`
- Shared parameter names: `refinement_factor` and `max_num_refinement_levels`
- `flat_multiscale` dispatches with simple `pattern` kwarg (no sub-dicts)
- `hierarchical` dispatches with `intra_level`/`inter_level` sub-dicts
- Backward compatibility: old parameter names (`mesh_node_distance`, `level_refinement_factor`, `max_num_levels`) still work with deprecation warnings
- Validates `grid_spacing` presence when `mesh_layout` is specified

**`flat.py`**
- `create_flat_multiscale_from_coordinates()` now accepts a simple `pattern="8-star"` argument instead of sub-dicts, matching Leif's API specification

**`archetype.py`**
- Updated `create_graphcast_graph` and `create_oskarsson_hierarchical_graph` to use new parameter names (`refinement_factor`, `max_num_refinement_levels`)

**`tests/test_mesh_layout.py`**
- 81 comprehensive tests covering coordinate creation, connectivity creation, API integration, backward compatibility, error handling, archetype equivalence, and 36 edge case tests

### API table (per specification in #78):

| `m2m_connectivity` | `mesh_layout` | `mesh_layout_kwargs` |
|---|---|---|
| `flat` | `rectilinear` | `pattern="8-star"` |
| `flat_multiscale` | `rectilinear` | `pattern="8-star"`, `refinement_factor=3`, `max_num_refinement_levels=None` |
| `hierarchical` | `rectilinear` | `intra_level={"pattern": "8-star"}`, `inter_level={"pattern": "knn", "k": 3}`, `refinement_factor=3`, `max_num_refinement_levels=None` |

### Motivation and context:
This lays the groundwork for supporting alternative mesh layouts (e.g., triangular in #80) by cleanly separating coordinate placement from connectivity logic through the `mesh_layout` abstraction.

### Dependencies:
No new dependencies required. Uses existing `networkx`, `numpy`, `scipy`.

## Issue Link

closes #78

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the documentation to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging)

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed* or *fixed*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

File: src/weather_model_graphs/create/base.py [hunk 1]
```diff
@@ -62,15 +78,23 @@ def create_all_graph_components(
     - "within_radius": Find all neighbours in grid within an absolute distance
         of `max_dist` or relative distance of `rel_max_dist` from each node in mesh
 
+    mesh_layout:
+    - "rectilinear": Regular rectilinear grid (default). Uses grid_spacing to
+        determine mesh node placement. Produces nodes with 4-star (cardinal) and
+        8-star (cardinal + diagonal) spatial adjacency edges.
```
Comment (leifdenby) [inline src/weather_model_graphs/create/base.py]: Can we be bit more precise thatn "*determine* mesh node placement" and say "Uniform regular grid with `grid_spacing` resolution" or something like that?

File: src/weather_model_graphs/create/base.py [hunk 2]
```diff
@@ -62,15 +78,23 @@ def create_all_graph_components(
...
+
+    mesh_layout_kwargs (for mesh_layout="rectilinear"):
+    - grid_spacing: float, distance between mesh nodes in coordinate units
+    - refinement_factor: int, refinement factor between levels (for multi-level)
```
Comment (leifdenby) [inline src/weather_model_graphs/create/base.py]: shouldn't this be `multi-level and hierarchical mesh graphs` rather than just `multi-level`?

File: src/weather_model_graphs/create/mesh/kinds/flat.py [hunk 3]
```diff
@@ -5,53 +5,62 @@
...
-    G_tot : networkx.Graph
-        The merged mesh graph
+    G_tot : networkx.DiGraph
+        The merged flat multiscale mesh graph
     """
-    # Check that level_refinement_factor is an odd integer
+
+    # Retrieve interlevel_refinement_factor from graph attributes
+    interlevel_refinement_factor = G_coords_list[0].graph.get(
+        "interlevel_refinement_factor", 3
```
Comment (leifdenby) [inline src/weather_model_graphs/create/mesh/kinds/flat.py]: rather than having a default here I think we should assert that this is set and raise an exception if not. I would also put in the docstring that this properties is expected to exist. Maybe we should start in the function by checking for the presence of all the required attributes? That way it is easy to see from the code what attributes will be required later

File: src/weather_model_graphs/create/mesh/kinds/flat.py [hunk 4]
```diff
@@ -5,53 +5,62 @@
 from .. import mesh as mesh_graph
 
 
-def create_flat_multiscale_mesh_graph(
-    xy, mesh_node_distance: float, level_refinement_factor: int, max_num_levels: int
+def create_flat_multiscale_from_coordinates(
+    G_coords_list,
```
Comment (leifdenby) [inline src/weather_model_graphs/create/mesh/kinds/flat.py]: add type annotation for calling args here

File: src/weather_model_graphs/create/mesh/kinds/flat.py
Comment (leifdenby) [inline src/weather_model_graphs/create/mesh/kinds/flat.py]: beautiful work on this module, love it!

File: src/weather_model_graphs/create/mesh/kinds/hierarchical.py [hunk 5]
```diff
@@ -6,62 +6,78 @@
...
-        Refinement factor between grid points and bottom level of mesh hierarchy
-    max_num_levels: int
-        The number of levels in the hierarchical mesh graph.
+    G_coords_list : list of networkx.Graph
+        List of undirected coordinate graphs, one per level. Each should have
+        nodes with "pos" and "type" attributes, and edges with "adjacency_type"
+        attributes. Created by create_multirange_2d_mesh_coordinates.
+    intra_level : dict or None
+        Intra-level connectivity options. Supports:
+        - pattern: str, "4-star" or "8-star" (default: "8-star")
```
Comment (leifdenby) [inline src/weather_model_graphs/create/mesh/kinds/hierarchical.py]: it would be good to describe again here what 4-star and 8-star means :)

File: src/weather_model_graphs/create/mesh/mesh.py [hunk 6]
```diff
@@ -56,71 +53,185 @@ def create_single_level_2d_mesh_graph(xy, nx, ny):
...
+    directed mesh graphs for flat_multiscale or hierarchical architectures.
 
     Parameters
     ----------
     max_num_levels : int
         Number of edge-distance levels in mesh graph
     xy : np.ndarray
         Grid point coordinates, shaped [N_grid_points, 2]
-    mesh_node_distance: float
+    grid_spacing : float
```
Comment (leifdenby) [inline src/weather_model_graphs/create/mesh/mesh.py]: reading this I am realising that `grid_spacing` is not a good name, sorry. That is because "grid" as a special meaning as-in nodes belonging to the `grid`. Maybe we should revert to `mesh_node_spacing`? That name does seem more applicable. Sorry for reverting this, I only just realised this now reading the docstring.

File: src/weather_model_graphs/create/mesh/mesh.py [hunk 7]
```diff
@@ -3,22 +3,19 @@
 from loguru import logger
 
 
-def create_single_level_2d_mesh_graph(xy, nx, ny):
+def create_single_level_2d_mesh_coordinates(xy, nx, ny):
```
Comment (leifdenby) [inline src/weather_model_graphs/create/mesh/mesh.py]: Maybe it would be better to call this `create_single_level_2d_mesh_primitive`? I think we need a name for the undirected graph that contains edges for the neighbourhood connectivity we might want to eventually build directed edges for. Maybe `mesh_blueprint` is better?

**Review by leifdenby — CHANGES_REQUESTED** [review]
This is really excellent! Thanks for doing this work. I have added some comments with suggestions for change. There are also some naming refactoring I think we should do (also in the comments). In general I would say this is 95% done though, well done! 🌟

**Review by leifdenby — CHANGES_REQUESTED** [review]
This is looking really good. In addressing my comments please make a separate commit for each comment you are addressing and paste a link to that commit in your reply. That makes it a lot easier for me to see how you addressed the comment.

In general I don't think we should have this pattern:

```python
def myfunction(arg=None):
    if arg is None:
        arg = 42
...
```

That hides the fact that the default for `arg` is actually `42`. Instead we should always put the default value in the call signature.

File: src/weather_model_graphs/create/base.py [hunk 8]
```diff
@@ -25,20 +25,85 @@
...
+    m2m_connectivity_kwargs : dict
+        Mutable dict of m2m connectivity keyword arguments.
+
+    Returns
+    -------
+    tuple[dict, dict]
+        Updated (mesh_layout_kwargs, m2m_connectivity_kwargs).
+    """
+    if "mesh_node_distance" in m2m_connectivity_kwargs and "mesh_node_spacing" not in mesh_layout_kwargs:
+        warnings.warn(
```
[the remark this answers is not in the document]
Comment (prajwal-tech07) [reply src/weather_model_graphs/create/base.py]: Addressed in [https://github.com/prajwal-tech07/weather-model-graphs/commit/e7e3c03](https://github.com/prajwal-tech07/weather-model-graphs/commit/e7e3c03) — switched _migrate_deprecated_kwargs to use logger.warning() from loguru instead of warnings.warn(). Also updated the backward-compat tests to capture loguru output in [https://github.com/prajwal-tech07/weather-model-graphs/commit/305e66a](https://github.com/prajwal-tech07/weather-model-graphs/commit/305e66a)

File: src/weather_model_graphs/create/mesh/coords.py [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+        g.graph["level"] = lev
+        # Store refinement factor so connectivity step can use it
+        g.graph["interlevel_refinement_factor"] = interlevel_refinement_factor
+        G_all_levels.append(g)
+
+    return G_all_levels
+
+
+def create_multirange_2d_mesh_graphs(
+    max_num_levels, xy, mesh_node_distance=3, level_refinement_factor=3
```
[the remark this answers is not in the document]
Comment (prajwal-tech07) [reply src/weather_model_graphs/create/mesh/coords.py]: Addressed in [https://github.com/prajwal-tech07/weather-model-graphs/commit/21d33e8](https://github.com/prajwal-tech07/weather-model-graphs/commit/21d33e8) — added type annotations to all parameters and also exposed the pattern argument to callers.

File: src/weather_model_graphs/create/mesh/coords.py [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+    G_coords_list = create_multirange_2d_mesh_primitives(
+        max_num_levels=max_num_levels,
+        xy=xy,
+        mesh_node_spacing=mesh_node_distance,
+        interlevel_refinement_factor=level_refinement_factor,
+    )
+
+    G_all_levels = []
+    for g_coords in G_coords_list:
+        g_directed = create_directed_mesh_graph(g_coords, pattern="8-star")
```
[the remark this answers is not in the document]
Comment (prajwal-tech07) [reply src/weather_model_graphs/create/mesh/coords.py]: Addressed in the same commit: [https://github.com/prajwal-tech07/weather-model-graphs/commit/21d33e8](https://github.com/prajwal-tech07/weather-model-graphs/commit/21d33e8) — create_multirange_2d_mesh_graphs now accepts pattern: str = "8-star" and passes it through to create_directed_mesh_graph.

**Comment by prajwal-tech07** [comment]
Thank you for the detailed review @leifdenby! I've addressed all your feedback with separate commits as requested. All 177 tests pass.

Also, whenever you get a chance, it would be great if you could review PRs #91 and #92 as well. Thanks!

**Review by leifdenby — CHANGES_REQUESTED** [review]
This is really getting there!

One general pattern we should avoid: All default parameter values (`pattern`, `refinement_factor`, etc) should not be set inline as follows:

```python
# anti-pattern, we should not do this
def my_function(refinement_factor):
    # do something...

refinement_factor = kwargs.get("refinement_factor", 42) # <-- this is bad, hiding default value inline rather than in function signature
my_function(refinement_factor=refinement_factor, ...)
```

Instead we should do this:

```python
def my_function(refinement_factor=42):
     # do something

my_function(**kwargs)
```

**Review by leifdenby — CHANGES_REQUESTED** [review]
Ok, now with this looking so clean I am starting seeing some structure appearing where a bit more refactoring would be nice. In particular, as I wrote in my comment I think we should do away with the "kinds" description for the meshes here in the code to create a clear separation between computing the mesh coordinates and creating the mesh connectivity.

**Comment by prajwal-tech07** [comment]
Hi Leif, I’ve addressed the requested changes and all checks are passing now. Could you please take another look and approve if everything looks good?

**Review by leifdenby — CHANGES_REQUESTED** [review]
This is looking good, just that refactor in `.create.base` remaining now.

To be clear: I am not suggesting that you have to add more helper functions, but you could if you want, e.g. something like `create_mesh_coordinates(layout=m2m_layout)` and `create_mesh_connectivity(connectivity=m2m_connectivity`).

What I am suggesting is that you reorder the if-statements. You should only have "Step 1", "Step 2" written once in the final implementation, so that "Step 1" produces a `G_mesh_coords` object, which can either be single `networkx.Graph` object or a `List[nx.Graph]` (add inline type annotation for this) and this object is then in Step 2 passed to the relevant function that applies the selected connectivity operation. That way we have proper separation of concerns so that steps are only written once but the functions used to process each step is dependent on the values of `mesh_layout` and `m2m_connectivity`.

Similarly, you should structure creating the kwargs for each so that they are grouped just above where they are used for either coordinate creation or connectivity creation. You could do this inside the two functions if you create them :) Separating the functionality out into two separate functions would reduce the size of `create_all_graph_components` which could be good.

**Comment by prajwal-tech07** [comment]
Hi @leifdenby! I've pushed the `base.py` refactoring you asked for in commit edbce82.

The `create_all_graph_components` function now has a clean two-step structure where each step is written exactly once:

- **Step 1** branches on `mesh_layout` to produce `G_mesh_coords: Union[nx.Graph, List[nx.Graph]]`
- **Step 2** branches on `m2m_connectivity` to create the directed mesh graph from those coordinates

This eliminates ~87 lines of duplicated coordinate-creation code that previously existed inside each of the 3 connectivity branches. The kwargs for each step are now grouped right above where they're used.

All 193 tests pass and all pre-commit hooks pass

File: src/weather_model_graphs/create/base.py [hunk 11]
```diff
@@ -124,29 +236,93 @@ def create_all_graph_components(
...
+            raise ValueError(
+                "mesh_layout='rectilinear' requires 'mesh_node_spacing' in "
+                "mesh_layout_kwargs (or 'mesh_node_distance' in "
+                "m2m_connectivity_kwargs for backward compatibility)."
+            )
+
+        if m2m_connectivity == "flat":
+            # Single-level mesh: compute nx/ny from spacing
+            range_x, range_y = np.ptp(xy, axis=0)
+            nx_mesh = int(range_x / mesh_node_spacing)
```
Comment (leifdenby) [inline src/weather_model_graphs/create/base.py]: I think the definition of `nx_mesh` and `ny_mesh` (and check of their values) should go inside of `create_single_level_2d_mesh_primitive(...)` since these variables are related to building the coordinate primitive

**Review by leifdenby — CHANGES_REQUESTED** [review]
This looks great with the last changes to `.create.base` - thank you!

Only one last very small tweak and this is ready to go in! Amazing work 🌟

**Comment by prajwal-tech07** [comment]
Addressed in `eb4533f` — moved the definition and validation of nx_mesh/ny_mesh from `base.py` into `create_single_level_2d_mesh_primitive() in `coords.py`.

The function now accepts an optional `mesh_node_spacing` keyword argument. When provided, `nx` and `ny` are computed internally from the coordinate extent of `xy`, along with the validation check. Existing callers that pass `nx`/`ny` directly (e.g. `create_multirange_2d_mesh_primitives`) continue to work unchanged.

All 194 tests pass. All pre-commit hooks pass.

**Review by leifdenby — CHANGES_REQUESTED** [review]
This looking really great! I'm reading through the current `weather_model_graphs.mesh.coords` and trying to decide if each function should be there or not. One thing I am thinking is that that module should only contain the functions that return `networkx.Graph` objects. Anything that returns `networkx.DiGraph` objects feels to me like it is representing connectivity. But I am not sure about this. What do you think?

Here's a quick diagram:

Image

`create.mesh.connectivity.hierarchical.create_hierarchical_from_coordinates()` for example calls `create.mesh.coords.create_directed_mesh_graph()` for example, and seems odd to me that `create_directed_mesh_graph()` should sit in `create.mesh.coords`.

One thing that would help me here would be if you could ensure all functions have type annotations for the return types. That way we can quickly scan through if `nx.Graph` or `nx.DiGraph` is returned

**Comment by prajwal-tech07** [comment]
Implemented the requested changes , Leif!!!

### 1. Move `create_directed_mesh_graph` to `connectivity/directed.py`

`create_directed_mesh_graph()` returns a `networkx.DiGraph`, so it belonged in `connectivity/`, not `coords.py` which is exclusively for `networkx.Graph` (coordinate layout) objects.

**New home:** `create/mesh/connectivity/directed.py`
**Backward-compatible:** still accessible via `from weather_model_graphs.create.mesh import coords; coords.create_directed_mesh_graph(...)` (re-exported with `noqa: F401`) and via `from weather_model_graphs.create.mesh.connectivity import create_directed_mesh_graph`

`flat.py` and `hierarchical.py` now import directly from `.directed` instead of routing through `mesh_coords`.

### 2. Return type annotations on all public functions

Added `-> networkx.Graph / networkx.DiGraph / List[...]` annotations to every public function across:
- `create/mesh/coords.py`
- `create/mesh/connectivity/directed.py`, `flat.py`, `hierarchical.py`
- `create/base.py`, `create/archetype.py`, `create/grid/grid.py`
- `networkx_utils.py`

### Tests

113 tests pass in `test_mesh_layout.py`, including 5 new test classes:
- `TestDirectedMeshGraphModuleLocation` — verifies the function is defined in `directed.py` and the re-exports are the same object
- `TestReturnTypeAnnotations` — verifies all functions have the correct `Graph` vs `DiGraph` annotations
- `TestGraphAttributeValidation` — edge cases for missing required graph attributes
- `TestDirectedMeshGraphAdditionalEdgeCases` — pattern filtering, graph attribute preservation, invalid pattern error messages

**Review by leifdenby — APPROVED** [review]
This look great now, well done with all the hard work @prajwal-tech07 :rocket:

I'll let you do the honours and merge this in :)

**Review by leifdenby — COMMENTED** [review]
I dont quite understand why you changed the way the plotting tests are done

File: tests/test_graph_creation.py
Comment (leifdenby) [inline tests/test_graph_creation.py:1]: I dont quite understand this - `tempfile.NamedTemporaryFile` should work in windows too, no?
